### PR TITLE
Promote LoadBalancerIPMode to GA

### PR DESCRIPTION
**Pull request by RyanAoh**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Promote LoadBalancerIPMode feature to GA.

#### Which issue(s) this PR fixes:
Part of https://github.com/kubernetes/enhancements/issues/1860

#### Does this PR introduce a user-facing change?
```release-note
LoadBalancerIPMode feature is now marked as GA.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding)
- [Usage](https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-ip-mode)

### CODE REVIEW

**Comment by aojea** [comment]
I always forget, if is locked to default don't we remove the features gates from the code too @thockin ?

**Comment by aojea** [comment]
I just realized there is no e2e test covering this feature, we should think in how to add one

**Comment by thockin** [comment]
e2e is challenging, indeed.

**Comment by thockin** [comment]
> I always forget, if is locked to default don't we remove the features gates from the code too @thockin?

We used to, but now with the multi-version compatibility, I think this is different - @jpbetz did we write up the new gate lifecycle?  When do we actually remove the gate usage from the code (not the gate, but the use of it to disable features) ?

**Comment by thockin** [comment]
I don't really know how to e2e this.  Would it be sufficient to test that iptables does not have a short-circuit rule?  That seems super brittle.

**Comment by aojea** [comment]
> I don't really know how to e2e this. Would it be sufficient to test that iptables does not have a short-circuit rule? That seems super brittle.

We already have unit and integration test https://github.com/search?q=repo%3Akubernetes%2Fkubernetes+LoadBalancerIPMode+path%3A%2F%5Epkg%5C%2Fproxy%5C%2F%2F&type=code and cloud-provider-kind implements this mode, so on my side I'm fine with the existing testing coverage for this feature

/lgtm

@thockin defer approval to you

**Comment by aojea** [comment]
/lgtm cancel

@RyanAoh please remove all the references to the feature gate from teh code base

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#ga-features

> When we set LockToDefault to true, we also remove all references to the feature gate from the codebase.

**Comment by liggitt** [comment]
> @RyanAoh please remove all the references to the feature gate from the code base

cc @Jefftree @jpbetz on normal process of graduating a gate removing all conditionals ... this means 1.32 would not be able to emulate <=1.31 where the feature was beta and disableable

**Comment by aojea** [comment]
> cc @Jefftree @jpbetz on normal process of graduating a gate removing all conditionals ... this means 1.32 would not be able to emulate <=1.31 where the feature was beta and disableable

ups, then we need to update the docs to reflect this

**Comment by Jefftree** [comment]
Thanks for catching this Jordan. We're in the process of updating the docs: https://github.com/kubernetes/community/issues/8082

**Comment by thockin** [comment]
So, IIUC, this is correct, and we can only remove the references in code after N+3, so v1.34.  The result being that in v1.34, we can still emulate 33, 32, and 31 (all of which will have been GA + locked).  Right?

**Comment by Jefftree** [comment]
> So, IIUC, this is correct, and we can only remove the references in code after N+3, so v1.35. The result being that in v1.35, we can still emulate 34, 33, and 32 (all of which will have been GA + locked). Right?
> 
> In the meantime, we might want to emulate v1.30 or v1.31, which was disableable.

Yep that is correct. v1.32 through v1.34 should all be able to emulate v1.31 which is disableable and the code references should stay for those releases. 

/lgtm
/hold cancel

**Comment by Jefftree** [comment]
For dealing with disablement tests with LockToDefault=true, an extra step is needed before disabling the feature gate in the tests:

```go
featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, utilfeature.DefaultFeatureGate, version.MustParse("1.31"))
```

See
https://github.com/kubernetes/kubernetes/pull/127093/files#diff-8868f5a770c851b2202de4360d037910dccf41dd67cfd213baf5e58eb8e4da88R451 and https://github.com/kubernetes/kubernetes/issues/127477.

cc @jpbetz

**Comment by Jefftree** [comment]
/hold

Test failures are legitimate related to feature disablement tests.

**Comment by aojea** [comment]
@RyanAoh once we have clarified the mechanics, it seems you need to do some changes to fix the failing tests 😄 

Sorry for the noise

**Comment by RyanAoh** [comment]
@aojea Got it, thanks for the heads-up! I'll fix the failing tests once we sort out the mechanics. No worries about the noise! 😄

**Comment by Jefftree** [comment]
@RyanAoh Can you rebase, I think there's some merge issues with the feature gate that's causing the unit test to fail.

File: pkg/features/versioned_kube_features.go [hunk 1]
```diff
@@ -634,6 +634,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	ServiceAccountTokenJTI: {
 		{Version: version.MustParse("1.29"), Default: false, PreRelease: featuregate.Alpha},
 		{Version: version.MustParse("1.30"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.32"), Default: true, PreRelease: featuregate.GA, LockToDefault: true},
```
Comment (Jefftree) [inline pkg/features/versioned_kube_features.go:637]: yeah, auto rebase changed the feature gate lol

Comment (Jefftree) [reply pkg/features/versioned_kube_features.go:637]: We recently made a change to add a line break between every feature which is what caused this, I don't think it should be a persistent problem with other PRs.

File: test/integration/service/loadbalancer_test.go [hunk 2]
```diff
@@ -707,6 +708,9 @@ func Test_ServiceLoadBalancerIPMode(t *testing.T) {
 
 	for _, tc := range testCases {
 		t.Run("", func(t *testing.T) {
+			if !tc.ipModeEnabled {
```
Comment (Jefftree) [inline test/integration/service/loadbalancer_test.go]:
```suggestion
			var testServerOptions *kubeapiservertesting.TestServerInstanceOptions
			if !tc.ipModeEnabled {
			  testServerOptions = &kubeapiservertesting.TestServerInstanceOptions{EmulationVersion: "1.31"}
```

File: test/integration/service/loadbalancer_test.go [hunk 3]
```diff
@@ -707,6 +708,9 @@ func Test_ServiceLoadBalancerIPMode(t *testing.T) {
...
+				featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, utilfeature.DefaultFeatureGate, version.MustParse("1.31"))
+			}
 			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.LoadBalancerIPMode, tc.ipModeEnabled)
 			server := kubeapiservertesting.StartTestServerOrDie(t, nil, framework.DefaultTestServerFlags(), framework.SharedEtcd())
```
Comment (Jefftree) [inline test/integration/service/loadbalancer_test.go:713]:
```suggestion
			server := kubeapiservertesting.StartTestServerOrDie(t, testServerOptions, framework.DefaultTestServerFlags(), framework.SharedEtcd())
```

**Pull request merged**